benchmarks/bench_ml.py

```diff
@@ -0,0 +1,422 @@
+
+                if not Algorithm in average_for_algo:
+                    average_for_algo[Algorithm] = []
+
+                objs = [[v[1] for v in d] for d in data] # leave only objective values
+                # here objs is a 2d list, where first dimension corresponds to
+                # particular repeat of experiment, and second dimension corresponds to index
+                # of optimization step during optimization
+                average_for_algo[Algorithm].append(objs)
+
+    # calculate averages
+    for Algorithm in average_for_algo:
+        mean_obj_vals = np.mean(average_for_algo[Algorithm], axis=0)
+        minimums = np.min(mean_obj_vals, axis=1)
+        f_calls = np.argmin(mean_obj_vals, axis=1)
+
+        min_mean = np.mean(minimums)
+        min_stdd = np.std(minimums)
+        min_best = np.min(minimums)
+
+        f_mean = np.mean(f_calls)
+        f_stdd = np.std(f_calls)
+        f_best = np.min(f_calls)
+
+        def fmt(float_value):
+            return ("%.3f" % float_value)
+
+        output = "|".join([fmt(min_mean) + " +/- " + fmt(min_stdd)] + [fmt(v) for v in [ min_best, f_mean, f_stdd, f_best]])
+        result = table_template + output
+        print("")
+        print(Algorithm)
+        print(result)
+
+
+def evaluate_optimizer(base_estimator, model, dataset, n_calls, seed_value):
+    """
+    Evaluates some estimator for the task of optimization of parameters of some
+    model, given limited number of model evaluations.
+    Parentheses on parameters are used to be able to run function with pool.map method.
+
+    :param base_estimator: Estimator to use for optimization.
+    :param model: str, name of the ML model class to be used for parameter tuning
+    :param dataset: str, name of dataset to train ML model on
+    :param n_calls: a budget of evaluations
+    :param seed_value: random seed, used to set the random number generator in numpy
+    :return: a list of paris (p, f(p)), where p is a dictionary of the form "param name":value,
+            and f(p) is performance measure value achieved by the model for configuration p.
+            Such list contains history of execution of optimization.
+    """
+
+
+    # below seed is necessary for processes which fork at the same time
+    # so that random numbers generated in processes are different
+    np.random.seed(seed_value)
+
+    problem = TesterClass(model, dataset)
+    space = problem.model_class[MODEL_PARAMETERS]
+
+    # initialization
+    estimator = base_estimator(random_state=seed_value)
+    dimensions = [v[0] for k, v in space.items()]
+    solver = Optimizer(dimensions, estimator, random_state=seed_value)
+
+    trace = []
+    best_y = np.inf
+    best_x = None
+
+    # optimization loop
+    for i in range(n_calls):
+        point_list = solver.ask()
+
+        point_dct = {k: v for k, v in zip(space.keys(), point_list)}  # convert list of dimension values to dictionary
```

Comment:
I think we should sort the keys of the space configuration both here and while specifying the dimensions to guarantee they are entered and received in the same order.